azurelinuxagent/ga/update.py

```diff
@@ -590,11 +593,31 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
 
             if self._processing_new_incarnation():
                 remote_access_handler.run()
+
+            if self._processing_new_extensions_goal_state():
+                try:
+                    UpdateHandler._cleanup_goal_state_history()
+                except Exception as exception:
+                    logger.warn("Error cleaning up the goal state history: {0}", ustr(exception))
+
         finally:
             if self._goal_state is not None:
                 self._last_incarnation = self._goal_state.incarnation
                 self._last_extensions_gs_id = self._goal_state.extensions_goal_state.id
 
+    @staticmethod
+    def _cleanup_goal_state_history():
+        archiver = StateArchiver(conf.get_lib_dir())
```

Comment:
This is only happening on an updated goal state, right? i.e. not every six seconds? 

Was initially a bit weary of creating a new StateArchiver every go around the loop, but I think this should be fine on the relatively sparse event of a new extensions goal state.

Reply:
the call is at line 599, protected by if self._processing_new_extensions_goal_state():

Reply:
(only on new goal state for extensions)